roachlint: add a pass to check for return of nil errors after checks

ajwerner opened the pull request:
This commit adds a lint pass which checks for if statments in functions
which return an error last which in their condition check if an error is
non-nil and then proceed to return a nil error.

Fixes #19316

Release note: None.

Commit:
roachlint: add a pass to check for return of nil errors after checks

This commit adds a lint pass which checks for if statments in functions
which return an error last which in their condition check if an error is
non-nil and then proceed to return a nil error.

Fixes #19316

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43264)
<!-- Reviewable:end -->

Commit:
*: mark benign nil error returns with nolint:returnerrcheck

This commit adds `nolint:returnerrcheck` annotations to all sites flagged
by the new returnerrcheck linter which seem intentional.

Release note: None.

Commit:
*: fix nil error returns caught by returnerrcheck

This commit includes fixes for cases caught by returnerrcheck. These changes
should be scrutinized.

Release note: None

ajwerner commented:
@bdarnell have any interest in reviewing or alternatively nominating a victim?

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

This is great, thanks for doing this!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-:-LwOnVap0KdIjQHNqjkR:bpt7lbc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/testutils/lint/passes/returnerrcheck/returnerrcheck.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOmgwF2ubfo6KMaCyU:-LwOmgwF2ubfo6KMaCyV:bb4bliy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/testutils/lint/passes/returnerrcheck/returnerrcheck.go#L82)):*
> ```Go
> 				pass.Report(analysis.Diagnostic{
> 					Pos:     n.Pos(),
> 					Message: fmt.Sprintf("unexpected nil error return after checking for a non-nil error"),
> ```

add something like `if this is not a mistake, add a "//nolint:returnerrcheck" comment` so someone who runs into this knows how to work around it
___
*[pkg/testutils/lint/passes/returnerrcheck/returnerrcheck.go, line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOn0eAEN1kIPs8NYIN:-LwOn0eAEN1kIPs8NYIO:blz5mxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/testutils/lint/passes/returnerrcheck/returnerrcheck.go#L90)):*
> ```Go
> }
> 
> func hasAcceptableAction(pass *analysis.Pass, errObj types.Object, statements []ast.Stmt) bool {
> ```

[nit] could use a comment


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Fantastic changes. I think we should be more careful with the handling of errors though.

> Reviewed 1 of 4 files at r1, 16 of 16 files at r2, 14 of 14 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-:-LwOsnxl7enrlYOzbXN0:b91z5fs)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/ccl/backupccl/backup.go, line 1042 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOqdd58qmNARxgHoew:-LwOqdd58qmNARxgHoex:b-f9lpeo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/ccl/backupccl/backup.go#L1042)):*
> ```Go
> 		defaultURI, urisByLocalityKV, err := getURIsByLocalityKV(to)
> 		if err != nil {
> 			return err
> ```

this will need a backport
___
*[pkg/cli/debug.go, line 870 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOqqjY2Zjt76LyVV-0:-LwOqqjZCxjSq5svIuLW:b-uwfdor) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cli/debug.go#L870)):*
> ```Go
> 				return err
> 			}
> 			return nil //nolint:returnerrcheck
> ```

This will read nicer as follows:
```go
if err == io.EOF {
   return nil
}
if err != nil {
   return err
}
```
___
*[pkg/cli/gen.go, line 121 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOqxdY9uHJoqUiQtVq:-LwOqxdY9uHJoqUiQtVr:b-kfh77h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cli/gen.go#L121)):*
> ```Go
> 	}
> 	if err != nil {
> 		return err
> ```

ditto backport
___
*[pkg/cmd/roachprod/install/cassandra.go, line 70 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOr2e17GAapZNYF-1U:-LwOr2e17GAapZNYF-1V:bjalces) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachprod/install/cassandra.go#L70)):*
> ```Go
> 				cmd := `nc -z $(hostname) 9042`
> 				if _, err := session.CombinedOutput(cmd); err != nil {
> 					return false, nil //nolint:returnerrcheck
> ```

maybe log the output if there was an error
___
*[pkg/cmd/roachtest/test_runner.go, line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrBse2QhGnPGz523x:-LwOrBse2QhGnPGz523y:b-mxsx3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachtest/test_runner.go#L510)):*
> ```Go
> 			// RunWithBuffer to toss the output.
> 			if _, err := c.RunWithBuffer(ctx, l, c.Node(node), "ls", perfArtifactsDir); err != nil {
> 				// perfArtifactsDir doesn't exist, nothing to fetch.
> ```

I would check that the error is indeed "does not exist" so as to catch other types of errors.
___
*[pkg/server/admin.go, line 1850 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrPLX-13VZfuGo1RK:-LwOrPLX-13VZfuGo1RL:b-72xx3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/server/admin.go#L1850)):*
> ```Go
> 	var repl *storage.Replica
> 	if err := s.server.node.stores.VisitStores(func(s *storage.Store) error {
> 		if r, _ := s.GetReplica(req.RangeID); r != nil {
> ```

Maybe only ignore the error if it was "no such range" and otherwise report the error - so that other error types bubble up properly.

Alternatively you could have an additional boolean return on `GetReplica` and use that to report "no such range". Then always bubble up `err` if non-nil.
___
*[pkg/server/server_systemlog_gc.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrWmk7zaWRPxDFPsX:-LwOrWmk7zaWRPxDFPsY:bnzqn73) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/server/server_systemlog_gc.go#L73)):*
> ```Go
> 	repl, err := s.node.stores.GetReplicaForRangeID(roachpb.RangeID(1))
> 	if err != nil {
> 		return timestampLowerBound, 0, nil //nolint:returnerrcheck
> ```

ditto my previous comment
___
*[pkg/sql/crdb_internal.go, line 2349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrhhgEffda_sdF-nf:-LwOrhhgEffda_sdF-ng:b-kfh77h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/crdb_internal.go#L2349)):*
> ```Go
> 		descriptors, err := getAllNodeDescriptors(p)
> 		if err != nil {
> 			return err
> ```

ditto backport
___
*[pkg/sql/descriptor.go, line 202 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrjOO8eiq8EZr2HNO:-LwOrjOO8eiq8EZr2HNP:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/descriptor.go#L202)):*
> ```Go
> 		}
> 		if err := table.MaybeFillInDescriptor(ctx, txn); err != nil {
> 			return err
> ```

ditto
___
*[pkg/sql/distsql_physical_planner.go, line 676 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrmdDAiNutpCE9Iou:-LwOrmdEAx5ax7lLDSWX:b-f0flso) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/distsql_physical_planner.go#L676)):*
> ```Go
> 		// written on startup, the most likely scenario is
> 		// that the node is ready. We therefore return no
> 		// error.
> ```

ditto previous - check the error type and only ignore for known errors
___
*[pkg/sql/physical_schema_accessors.go, line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrrI52y3u7YgeKiq-:-LwOrrI52y3u7YgeKiq0:b-kfh77h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/physical_schema_accessors.go#L197)):*
> ```Go
> 	ok, schemaID, err := a.IsValidSchema(ctx, txn, dbID, name.Schema())
> 	if err != nil {
> 		return nil, err
> ```

ditto backport
___
*[pkg/sql/execinfrapb/flow_diagram.go, line 726 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrtnyCK57M8VabctK:-LwOrtnyCK57M8VabctL:b-kfh77h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/execinfrapb/flow_diagram.go#L726)):*
> ```Go
> 	d, err := GeneratePlanDiagram(sql, flows, showInputTypes)
> 	if err != nil {
> 		return "", url.URL{}, err
> ```

ditto backport
___
*[pkg/sql/flowinfra/outbox.go, line 285 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrvL35O4PtcF5QJmW:-LwOrvL35O4PtcF5QJmX:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/flowinfra/outbox.go#L285)):*
> ```Go
> 					err := m.flush(ctx)
> 					if err != nil {
> 						return err
> ```

ditto
___
*[pkg/sql/opt/exec/execbuilder/scalar.go, line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOs-_JDTZlEhhBG4YY:-LwOs-_JDTZlEhhBG4YZ:b-jnczkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/opt/exec/execbuilder/scalar.go#L101)):*
> ```Go
> 				// during execution where they are correctly handled. Note that in some
> 				// cases we might not even get an error (if this particular expression
> 				// does not get evaluated when the query runs, e.g. it's inside a CASE).
> ```

check that the error is not an assertion before ignoring it. Let assertions bubble up.
___
*[pkg/sql/opt/exec/execbuilder/scalar.go, line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOs5O757z5D50GNIE0:-LwOs5O757z5D50GNIE1:b-gzil0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/opt/exec/execbuilder/scalar.go#L110)):*
> ```Go
> 				newExpr, err = tree.ReType(tree.DNull, texpr.ResolvedType())
> 				if err != nil {
> 					return texpr, nil //nolint:returnerrcheck
> ```

Wait why is this ignored?
___
*[pkg/storage/cloud/external_storage.go, line 399 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsFJV3thvuTfRT409:-LwOsFJV3thvuTfRT40A:b-kfh77h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/cloud/external_storage.go#L399)):*
> ```Go
> 	stat, err := l.blobClient.Stat(ctx, joinRelativePath(l.base, basename))
> 	if err != nil {
> 		return 0, err
> ```

ditto backport
___
*[pkg/storage/engine/merge.go, line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsHM_52wUrl9U6gBu:-LwOsHM_52wUrl9U6gBv:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/merge.go#L69)):*
> ```Go
> 	srcBytes, err := serializeMergeInputs(sources...)
> 	if err != nil {
> 		return roachpb.InternalTimeSeriesData{}, err
> ```

ditto
___
*[pkg/storage/engine/pebble.go, line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsJVL4J3m-NQ_uK_V:-LwOsJVL4J3m-NQ_uK_W:bz2d7wq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/pebble.go#L177)):*
> ```Go
> 			// min/max timestamp properties. This mimics the behavior of
> 			// TimeBoundTblPropCollector.
> 			return nil //nolint:returnerrcheck
> ```

Check the error type and only ignore on known errors.
___
*[pkg/storage/engine/row_counter.go, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsNvd2GRAdWlrki07:-LwOsNvd2GRAdWlrki08:b-m9fe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/row_counter.go#L39)):*
> ```Go
> 	row, err := keys.EnsureSafeSplitKey(key)
> 	if err != nil || len(key) == len(row) {
> 		return nil //nolint:returnerrcheck
> ```

Check the error type.
___
*[pkg/testutils/reduce/reduce.go, line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsXnNBuNdB5FCHSe6:-LwOsXnNBuNdB5FCHSe7:bftp48v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/testutils/reduce/reduce.go#L224)):*
> ```Go
> 			next, err := findNextInteresting(vs)
> 			if err != nil {
> 				//nolint:returnerrcheck
> ```

Is this meant to ignore all errors or just io.EOF? If the latter then be specific.
___
*[pkg/util/json/contains.go, line 205 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsbEXCQq6HIbF0tiR:-LwOsbEXCQq6HIbF0tiS:b-kfh77h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/util/json/contains.go#L205)):*
> ```Go
> 		objectsMatch, err := quadraticJSONArrayContains(j.objects, contained.objects)
> 		if err != nil {
> 			return false, err
> ```

ditto backport
___
*[pkg/util/log/file.go, line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOskSc0ta00XMaFH9Z:-LwOskSc0ta00XMaFH9_:bgaw0cm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/util/log/file.go#L283)):*
> ```Go
> 	dir, err := l.logDir.get()
> 	if err != nil {
> 		// No log directory configured: simply indicate that there are no
> ```

Maybe replace the error return by a boolean and check that.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
I'll make a backport pass through the error fixes afterwards.

@knz see how you feel about the steps I took to address the other comments. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-:-LwQ-kcz-57QW5GTuxn2:b-2ppntx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @knz, and @mjibson)

___
*[pkg/cli/debug.go, line 870 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOqqjY2Zjt76LyVV-0:-LwPCRTUA5rUFQ_nkXzS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cli/debug.go#L870)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This will read nicer as follows:
```go
if err == io.EOF {
   return nil
}
if err != nil {
   return err
}
```
</blockquote></details>

Done.
___
*[pkg/cmd/roachprod/install/cassandra.go, line 70 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOr2e17GAapZNYF-1U:-LwPCvaw3mH9jQWkEE66:bx6zld6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachprod/install/cassandra.go#L70)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

maybe log the output if there was an error
</blockquote></details>

The common case here is going to be `exit status 1` until the cassandra process starts listening on the port. I think logging would just create noise. 
___
*[pkg/cmd/roachtest/test_runner.go, line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrBse2QhGnPGz523x:-LwPDXep0ocH4DfbxtSN:bwt994f) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachtest/test_runner.go#L510)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would check that the error is indeed "does not exist" so as to catch other types of errors.
</blockquote></details>

Done. The string parsing is a bit brittle but the relevant exit code from ls (`2`) doesn't make it through roachprod. 
___
*[pkg/server/admin.go, line 1850 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrPLX-13VZfuGo1RK:-LwQ7-NI68atfdrqPzSn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/server/admin.go#L1850)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Maybe only ignore the error if it was "no such range" and otherwise report the error - so that other error types bubble up properly.

Alternatively you could have an additional boolean return on `GetReplica` and use that to report "no such range". Then always bubble up `err` if non-nil.
</blockquote></details>

Done.
___
*[pkg/server/server_systemlog_gc.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrWmk7zaWRPxDFPsX:-LwQ73JW9EmGhhnPJb90:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/server/server_systemlog_gc.go#L73)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto my previous comment
</blockquote></details>

Done.
___
*[pkg/sql/distsql_physical_planner.go, line 676 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrmdDAiNutpCE9Iou:-LwPwt71Aczn6SH5GDqG:b-pn24ck) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/distsql_physical_planner.go#L676)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto previous - check the error type and only ignore for known errors
</blockquote></details>

I've added a TODO on this one because I don't know which errors are expected. 
___
*[pkg/sql/opt/exec/execbuilder/scalar.go, line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOs-_JDTZlEhhBG4YY:-LwQ-aC9-kuzxPCG2LQb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/opt/exec/execbuilder/scalar.go#L101)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

check that the error is not an assertion before ignoring it. Let assertions bubble up.
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/execbuilder/scalar.go, line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOs5O757z5D50GNIE0:-LwPxDjbEyswS3QntjbR:bnx3utx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/opt/exec/execbuilder/scalar.go#L110)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Wait why is this ignored?
</blockquote></details>

Good question. I didn't have confidence on this one. I've added a TODO. If you or any other passerby knows that we should not swallow this error I'm happy to return it. 
___
*[pkg/storage/cloud/external_storage.go, line 399 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsFJV3thvuTfRT409:-LwPzp_P8B0HP1nRdKP9:b-2k82ta) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/cloud/external_storage.go#L399)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto backport
</blockquote></details>

This is new code in 20.1
___
*[pkg/storage/engine/merge.go, line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsHM_52wUrl9U6gBu:-LwPzt2P3U7yiW1rwYTn:b-8dud7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/merge.go#L69)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

This is new code in 20.1 
___
*[pkg/storage/engine/pebble.go, line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsJVL4J3m-NQ_uK_V:-LwPxYHy4LqouUr2PNWk:bfaku6s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/pebble.go#L177)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Check the error type and only ignore on known errors.
</blockquote></details>

@petermattis/@itsbilal/@ajkr have thoughts on what errors we might expect? I'm okay just leaving this for now if y'all are. 
___
*[pkg/storage/engine/row_counter.go, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsNvd2GRAdWlrki07:-LwPxp1_Ep8T3NS59uYb:b67hw6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/row_counter.go#L39)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Check the error type.
</blockquote></details>

It's not clear to me which errors I would or would not want to ignore. I went and read the implementation of this function and there's a lot of different errors which could be returned and I couldn't come up with a good rule. I added a TODO. 
___
*[pkg/testutils/lint/passes/returnerrcheck/returnerrcheck.go, line 82 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOmgwF2ubfo6KMaCyU:-LwPzZK65R-Fnl-CX89l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/testutils/lint/passes/returnerrcheck/returnerrcheck.go#L82)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

add something like `if this is not a mistake, add a "//nolint:returnerrcheck" comment` so someone who runs into this knows how to work around it
</blockquote></details>

Done.
___
*[pkg/testutils/lint/passes/returnerrcheck/returnerrcheck.go, line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOn0eAEN1kIPs8NYIN:-LwPz_4e8OJ0vwpeEn7s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/testutils/lint/passes/returnerrcheck/returnerrcheck.go#L90)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] could use a comment
</blockquote></details>

Done.
___
*[pkg/testutils/reduce/reduce.go, line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsXnNBuNdB5FCHSe6:-LwPzcw68Od8lDwHxS6P:b-nnyfxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/testutils/reduce/reduce.go#L224)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Is this meant to ignore all errors or just io.EOF? If the latter then be specific.
</blockquote></details>

I have no idea. It's just used in testing so I'm not that worried about it. @mjibson ?
___
*[pkg/util/log/file.go, line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOskSc0ta00XMaFH9Z:-LwQ7Q2f3ZNcoR_TgGDA:btupv1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/util/log/file.go#L283)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Maybe replace the error return by a boolean and check that.
</blockquote></details>

:man_shrugging: 


<!-- Sent from Reviewable.io -->

madelynnblue reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @knz)

___
*[pkg/testutils/reduce/reduce.go, line 224 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsXnNBuNdB5FCHSe6:-LwQ8z4b5IrXx76qkZ7v:bbi4j2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/testutils/reduce/reduce.go#L224)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I have no idea. It's just used in testing so I'm not that worried about it. @mjibson ?
</blockquote></details>

This is meant to ignore all errors. Adding the nolint comment is the correct thing here.


<!-- Sent from Reviewable.io -->

mjibson commented:
Also yay for another linter using the go/analysis stuff.

knz reviewed (commented):
> Reviewed 2 of 4 files at r1, 1 of 30 files at r4, 18 of 18 files at r5, 12 of 12 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[pkg/cmd/roachprod/install/cassandra.go, line 70 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOr2e17GAapZNYF-1U:-LwVJ4e46W71RMH5FmGf:bftgm30) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachprod/install/cassandra.go#L70)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The common case here is going to be `exit status 1` until the cassandra process starts listening on the port. I think logging would just create noise. 
</blockquote></details>

Thanks for explaining. Please add a comment with your sentence above.
___
*[pkg/cmd/roachtest/test_runner.go, line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrBse2QhGnPGz523x:-LwVJLm95OUAAoq6O2oP:b-v23tt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachtest/test_runner.go#L510)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Done. The string parsing is a bit brittle but the relevant exit code from ls (`2`) doesn't make it through roachprod. 
</blockquote></details>

A different approach: either
- if RunWithBuffer could be instructed to only capture stdout,
- otherwise, run the command `"env", "d="+perfActifactsDir, "bash", "-c", "ls \"$d\" 2>/dev/null"`
then check that stdout is empty.
in any case, check that the error returned by RunWithBuffer is an instance of os.ExitError. Anything else would need to be reported as failure.
___
*[pkg/sql/distsql_physical_planner.go, line 676 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrmdDAiNutpCE9Iou:-LwVKa-BD-XaeYZNtm_i:b-2sqgsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/distsql_physical_planner.go#L676)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I've added a TODO on this one because I don't know which errors are expected. 
</blockquote></details>

It's `errors.IsType(err, (*gossip.KeyNotPresentError)(nil))` (and maybe add a predicate function in the `gossip` package.)

The other error that can occur and which we definitely do not want to ignore is proto.DecodeError (gossip payload was corrupted).
___
*[pkg/sql/opt/exec/execbuilder/scalar.go, line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOs5O757z5D50GNIE0:-LwVLDWUERYLKf868QhZ:b-cyvlio) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/opt/exec/execbuilder/scalar.go#L110)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Good question. I didn't have confidence on this one. I've added a TODO. If you or any other passerby knows that we should not swallow this error I'm happy to return it. 
</blockquote></details>

I went and checked: the function `ReType` never returns an error. I think it would be good in this case to remove the error return altogether and simplify this code here. 
___
*[pkg/storage/engine/pebble.go, line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsJVL4J3m-NQ_uK_V:-LwVLt3O1TjTK-FpK9mi:ba9h54s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/pebble.go#L177)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

@petermattis/@itsbilal/@ajkr have thoughts on what errors we might expect? I'm okay just leaving this for now if y'all are. 
</blockquote></details>

I'm also surprised that this code reutrns nil and not the `err` object. The comment just above says "the last record was an intent". If that is true and the unmarshal fail, that means there's corruption. That does not look like a situation to paper over.
___
*[pkg/storage/engine/row_counter.go, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsNvd2GRAdWlrki07:-LwVMK8A58C1QUeoyvqH:bxwri11) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/row_counter.go#L39)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It's not clear to me which errors I would or would not want to ignore. I went and read the implementation of this function and there's a lot of different errors which could be returned and I couldn't come up with a good rule. I added a TODO. 
</blockquote></details>

I also went and looked and this is just a case of a mis-designed interface.
The function `EnsureSafeSplitKey` assumes the input is a SQL key already, whereas the caller here doesn't even know this.

Suppose we had a separate function `IsValidSQLKey(roachpb.Key) bool`.

Then the code here would be:
```go
if !keys.IsValidSQLKey(key) {
   return nil
}
row, err := keys.EnsureSafeSplitKey(key)
if err != nil {
   return err
}
```

Meanwhile, it so happens that our current definition for this `IsValidSQLKey` happens to be "does this key decode properly" so it looks like we can share the code. But the caller cannot assume to know this! So we need an API that does both. What about:

```go
package keys

func CheckIsSQLKeyAndEnsureSplit(
   key roachpb.Key,
) (isValidSQLKey bool, rowPrefixLen int, err error) {
  row, err := EnsureSafeSplitKey(key)
  // Currently it so happens that a failure to decode is exactly
  // equivalent to the key not being valid SQL.
  return err == nil, row, nil
}
```
(I would even specialize the function `EnsureSafeSplitKey` to avoid constructing error objects on the heap to throw them away immediately!)


Then here in the caller:
```go
ok, row, err := keys.CheckIsSQLKeyAndEnsureSplit(key)
if err != nil {
   return err
}
if !ok {
  return nil
}
```
___
*[pkg/util/log/file.go, line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOskSc0ta00XMaFH9Z:-LwVP1QXDkhi-SiHSpi6:b-oh3ezv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/util/log/file.go#L283)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

:man_shrugging: 
</blockquote></details>

I don't understand?


<!-- Sent from Reviewable.io -->

Commit:
sql/sem/tree: remove error from tree.ReType

ReType never returns an error so let's remove the error from its signature
and simplify the logic of its callers.

Release note: None

ajwerner reviewed (commented):
@knz I addressed the comments I felt comfortable addressing for this pass. I'd prefer to merge this and address remaining concerns in a follow up. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-:-Lw_6zQ60RLw1-Wey05I:bxxdnig)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @knz)

___
*[pkg/cmd/roachprod/install/cassandra.go, line 70 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOr2e17GAapZNYF-1U:-Lw_6qrt-yXBh6rWBQfl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachprod/install/cassandra.go#L70)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Thanks for explaining. Please add a comment with your sentence above.
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/test_runner.go, line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrBse2QhGnPGz523x:-LwVQG7Z1ALIIk0V0Oxs:b-v0hg5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachtest/test_runner.go#L510)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

A different approach: either
- if RunWithBuffer could be instructed to only capture stdout,
- otherwise, run the command `"env", "d="+perfActifactsDir, "bash", "-c", "ls \"$d\" 2>/dev/null"`
then check that stdout is empty.
in any case, check that the error returned by RunWithBuffer is an instance of os.ExitError. Anything else would need to be reported as failure.
</blockquote></details>

See how you feel about this. It should be less brittle as it doesn't rely on `ls`'s output. 
___
*[pkg/sql/opt/exec/execbuilder/scalar.go, line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOs5O757z5D50GNIE0:-Lw_0KapD9htwY3DyyGa:bbvbquc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/opt/exec/execbuilder/scalar.go#L110)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I went and checked: the function `ReType` never returns an error. I think it would be good in this case to remove the error return altogether and simplify this code here. 
</blockquote></details>

Done. Added another commit to make this change. 
___
*[pkg/storage/engine/pebble.go, line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsJVL4J3m-NQ_uK_V:-Lw_6JIz3Dm_2NeOsNaG:b8kxygn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/pebble.go#L177)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I'm also surprised that this code reutrns nil and not the `err` object. The comment just above says "the last record was an intent". If that is true and the unmarshal fail, that means there's corruption. That does not look like a situation to paper over.
</blockquote></details>

I'm still waiting for input from the storage team here. In the meantime, not changing the behavior at least does not pose risks. 
___
*[pkg/storage/engine/row_counter.go, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsNvd2GRAdWlrki07:-Lw_4aBhADGaE1YbhqX3:bx272ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/row_counter.go#L39)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I also went and looked and this is just a case of a mis-designed interface.
The function `EnsureSafeSplitKey` assumes the input is a SQL key already, whereas the caller here doesn't even know this.

Suppose we had a separate function `IsValidSQLKey(roachpb.Key) bool`.

Then the code here would be:
```go
if !keys.IsValidSQLKey(key) {
   return nil
}
row, err := keys.EnsureSafeSplitKey(key)
if err != nil {
   return err
}
```

Meanwhile, it so happens that our current definition for this `IsValidSQLKey` happens to be "does this key decode properly" so it looks like we can share the code. But the caller cannot assume to know this! So we need an API that does both. What about:

```go
package keys

func CheckIsSQLKeyAndEnsureSplit(
   key roachpb.Key,
) (isValidSQLKey bool, rowPrefixLen int, err error) {
  row, err := EnsureSafeSplitKey(key)
  // Currently it so happens that a failure to decode is exactly
  // equivalent to the key not being valid SQL.
  return err == nil, row, nil
}
```
(I would even specialize the function `EnsureSafeSplitKey` to avoid constructing error objects on the heap to throw them away immediately!)


Then here in the caller:
```go
ok, row, err := keys.CheckIsSQLKeyAndEnsureSplit(key)
if err != nil {
   return err
}
if !ok {
  return nil
}
```

</blockquote></details>

I started to do this and have backed off. I attempted to do it especially with an eye towards avoiding the allocation of the errors. Avoiding the allocations underneath `encoding.DecodeUvarintAscending` started to feel onerous. I'm not that comfortable making the interface change you suggest. I'm happy to add a TODO and/or create an issue which we can revisit in a later PR. Sorry I'd rather not sit on this for too much longer. 
___
*[pkg/util/log/file.go, line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOskSc0ta00XMaFH9Z:-Lw_1sei3Fh8yrEcW3wa:b1nugpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/util/log/file.go#L283)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't understand?
</blockquote></details>

I was attempting to communicate the I didn't have a strong feeling but would prefer to avoid changing the code. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @knz)

___
*[pkg/storage/engine/pebble.go, line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsJVL4J3m-NQ_uK_V:-Lw_BKM53eQ-K8UNVShr:b-la7d2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/pebble.go#L177)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm still waiting for input from the storage team here. In the meantime, not changing the behavior at least does not pose risks. 
</blockquote></details>

This LGTM. Note the comment that the behavior here mimics the behavior of the C++ `TimeBoundTblPropCollector`. We shouldn't change one without changing the other. I'd be fine for there to be an issue tracking this.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @knz)

___
*[pkg/storage/engine/pebble.go, line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsJVL4J3m-NQ_uK_V:-Lw_D4p_-7EXgyfmByHN:bkpavng) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/pebble.go#L177)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This LGTM. Note the comment that the behavior here mimics the behavior of the C++ `TimeBoundTblPropCollector`. We shouldn't change one without changing the other. I'd be fine for there to be an issue tracking this.
</blockquote></details>

Added a TODO(petermattis) and created https://github.com/cockroachdb/cockroach/issues/43422
___
*[pkg/storage/engine/row_counter.go, line 39 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsNvd2GRAdWlrki07:-Lw_FGlf5UURvyOcZyWC:b-4mq490) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/row_counter.go#L39)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I started to do this and have backed off. I attempted to do it especially with an eye towards avoiding the allocation of the errors. Avoiding the allocations underneath `encoding.DecodeUvarintAscending` started to feel onerous. I'm not that comfortable making the interface change you suggest. I'm happy to add a TODO and/or create an issue which we can revisit in a later PR. Sorry I'd rather not sit on this for too much longer. 
</blockquote></details>

Added the TODO and issue. 


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @knz)

___
*[pkg/storage/engine/pebble.go, line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOsJVL4J3m-NQ_uK_V:-Lw_NNvNCs8F0XaVd6gH:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/storage/engine/pebble.go#L177)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Added a TODO(petermattis) and created https://github.com/cockroachdb/cockroach/issues/43422
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 16 files at r7, 5 of 5 files at r8, 12 of 12 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-:-LwcTpCXC8fxQ-MC-oZ7:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[pkg/cmd/roachtest/test_runner.go, line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrBse2QhGnPGz523x:-LwcS8AG8mgr-zRmwTHf:b-gdhiju) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachtest/test_runner.go#L510)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

See how you feel about this. It should be less brittle as it doesn't rely on `ls`'s output. 
</blockquote></details>

IMHO "ls returns no output on its stdout" is a stronger signal that the directory is not to be investigated further.
Your current approach is one step better than the original code, but still vulnerable to the existence of a file whose name contains `doesNotExistStr`. 
I'll let you decide what you want to do about it.
___
*[pkg/util/log/file.go, line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOskSc0ta00XMaFH9Z:-LwcTj9P1MX-WfChNxuz:b-4aqn3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/util/log/file.go#L283)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I was attempting to communicate the I didn't have a strong feeling but would prefer to avoid changing the code. 
</blockquote></details>

Then please leave a TODO with my name on it. I'll pick it up.


<!-- Sent from Reviewable.io -->

Commit:
util/log: make logDir.get return a boolean instead of an error

The error, if returned, only indicated that the directory was not set.
This commit changes to API to return a boolean. Callers can then return the
error if they so choose.

Release note: None

ajwerner reviewed (commented):
TFTR! I'll follow up with backports.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-:-LwnluoY3ebSJtdYaFuq:b-rjinu6)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner and @knz)

___
*[pkg/cmd/roachtest/test_runner.go, line 510 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrBse2QhGnPGz523x:-LwnLlnw36zFHEt_pKsO:b6xgt69) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/cmd/roachtest/test_runner.go#L510)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

IMHO "ls returns no output on its stdout" is a stronger signal that the directory is not to be investigated further.
Your current approach is one step better than the original code, but still vulnerable to the existence of a file whose name contains `doesNotExistStr`. 
I'll let you decide what you want to do about it.
</blockquote></details>

Alright I think I've got something good now. It will return without an error if the directory does not exist or if it does. It will return an error if there exists a non-directory entity in the filesystem at that path or any other sort of error. 
___
*[pkg/sql/distsql_physical_planner.go, line 676 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOrmdDAiNutpCE9Iou:-LwnXlW1D9_EQD0FroLm:bl4uncv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/sql/distsql_physical_planner.go#L676)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

It's `errors.IsType(err, (*gossip.KeyNotPresentError)(nil))` (and maybe add a predicate function in the `gossip` package.)

The other error that can occur and which we definitely do not want to ignore is proto.DecodeError (gossip payload was corrupted).
</blockquote></details>

I'll do this in a follow-up to address the TODO. 
___
*[pkg/util/log/file.go, line 283 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/43264#-LwOskSc0ta00XMaFH9Z:-LwnRGhxEuJbuvRmTATk:b-fmwijh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a92365204a714f2e53da5684ada17deaae0ce7ce/pkg/util/log/file.go#L283)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Then please leave a TODO with my name on it. I'll pick it up.
</blockquote></details>

Done. I addressed the comment because it was easier than I had anticipated. Sorry for not digging deeper on the first pass. 


<!-- Sent from Reviewable.io -->
